Add an exit syscall (was: Add a syscall to notify the kernel that the app panicked)

torfmaster opened the pull request:
### Pull Request Overview

This pull request is an attempt to address https://github.com/tock/libtock-rs/issues/133 by adding a syscall to allow an app to notify the kernel that an app has panicked. ~The app will not be resumed anymore neither will there be callbacks scheduled~.  The app will be stopped/restarted depending on the first on only parameter of the syscall (0=stop, 1=restart).


### Testing Strategy

This pull request was tested by flashing the `panic` example of https://github.com/torfmaster/libtock-rs/tree/feature/panic-syscall to verify that the syscall is triggered and the app is not resumed.

### TODO or Help Wanted

The implementation is quite raw and there are some questions to be clarified.
Open questions are:
 
 * How to handle panicked apps? I'm in favour of letting the app die and maybe even let the kernel panic because no possibility to recover an app from a "panicked" state than restarting the whole system
 * ~Do we need parameters? I find it hard to squeeze the cause for a panic into some bytes so I would not pass parameters to the syscall~

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

torfmaster commented:
You understand that correctly. We could use the restart mechanism. 
Although, I think from a security perspective it is a not a good idea to give an attacker the possibility of infinitely often crash and restart the app with completely deterministic memory layout. So I tend to leave the app in a panicked state or even panic the whole kernel.

I see four possibilities here:
 
 * restart panicked app -> try to reuse the `FaultResponse` mechanism.
 * stop panicked app -> also try to reuse the mechanism
 * panic the kernel
 * let the abi consumer choose (i.e. giving the syscall a parameter which is mapped to a `FaultResponse`)

Any opinions?

torfmaster commented:
I pushed a commit where I stop the app up "notify_panic" syscall. However, I am open for any other solution.

However, the problem of pending interrupts remains unsolved. As far as I can judge the capsules don't track which apps enable which interrupts so we can't disable them.

bradjc commented:
This is similar to #1559, which advocates for some way for processes to signal to the kernel that they shouldn't be scheduled anymore.

The `FaultResponse` mechanism should work, although there are a series of outstanding PRs to make restarting apps work correctly.

I also support apps being able to tell the kernel "i'm done, do not bother to run me anymore" or "i crashed, but please restart me". I don't think apps should be able to lead to the kernel panic/crashing in any case, but that is really a separate discussion.

I think we should call this syscall `exit` to make it more general.

torfmaster commented:
Okay, I will call the syscall `exit` and give a parameter

 * 0 means stop
 * 1 means restart

We can still add panic functionality later.

gendx commented:
As a comparison point, we stumbled across this problem in developing OpenSK, and a "hack" (see our [patch](https://github.com/google/OpenSK/blob/master/patches/libtock-rs/02-panic_console.patch)) was to modify the panic handler in userspace to:
1. write the panic message via the `Console` driver,
1. write to memory address 0, triggering the MPU to detect a memory fault (see in particular [here](https://github.com/google/OpenSK/blob/master/patches/libtock-rs/02-panic_console.patch#L58-L62)).

The memory fault then causes a kernel panic, which is useful in showing the app status, but of course also blocks everything until the board is reset.

https://github.com/tock/tock/blob/c6ad079df7858f4da1a032a052922322ebf21ed2/kernel/src/process.rs#L662-L665

Some kind of `exit` syscall that triggers printing the app status but without panicking the whole kernel would be useful.

And related to that, I now realize that an app shouldn't cause a kernel panic due to a memory fault, this should instead cause the kernel to print the diagnostic but then either stop or restart this app (depending on some pre-defined policy), without affecting other apps.

bradjc commented:
> And related to that, I now realize that an app shouldn't cause a kernel panic due to a memory fault, this should instead cause the kernel to print the diagnostic but then either stop or restart this app (depending on some pre-defined policy), without affecting other apps.

With #1565 it is now possible (I think) to implement exactly that.

labbott commented:
What I'd really like to see is a reliable way to get the panic message from userspace so that if the console driver itself isn't working we can still get the message. This may be bring up specific but robust error reporting is important.

torfmaster commented:
> What I'd really like to see is a reliable way to get the panic message from userspace so that if the console driver itself isn't working we can still get the message. This may be bring up specific but robust error reporting is important.

In principle you got do a allow on the `core::panic::PanicInfo` payload and dump a few last words using the "debug" capsule before exiting. But I thinks that a different subject.

torfmaster commented:
I added an exit syscall and rebased on current master.

gendx commented:
> What I'd really like to see is a reliable way to get the panic message from userspace so that if the console driver itself isn't working we can still get the message. This may be bring up specific but robust error reporting is important.

I definitely agree. Using the existing console driver was the easiest to implement and worked reasonably well for us, but a more robust panic-specific method would be useful.

torfmaster commented:
> I definitely agree. Using the existing console driver was the easiest to implement and worked reasonably well for us, but a more robust panic-specific method would be useful.

I agree we should provide some possibility for panic_handlers to provide important information. However, the first goal of this PR is to fix https://github.com/tock/libtock-rs/issues/133 to avoid UD after panicking. Shall we create a separate issue "output panic cause on panic of user space apps" for this? I would be happy to assist on solving this in a separate PR.

hudson-ayers commented on kernel/src/process.rs:
```diff
@@ -375,6 +379,9 @@ pub enum State {
     /// of its state is reset as if it has not been executed yet.
     Unstarted,
+
+    /// The process has a panic and aborted.
```
Given that we are discussing using this syscall for cases other than panics in libtock-rs, I suggest something more like "The process has signaled it is permanently finished."

hudson-ayers commented on kernel/src/process.rs:
```diff
@@ -1605,4 +1531,100 @@ impl<C: 'static + Chip> Process<'a, C> {
+            // `FaultResponse::Stop` policy.
+            if !restart_policy.should_restart(self) {
+                return;
+            }
```
If `restart_app()` is called in response to a fault and then `restart_policy.should_restart()` indicates that the app should not in fact restart, the process will now be left in the `State::Exited` state when before it would have been left in the `State::StoppedFaulted` state. To prevent this bug, I think `restart_app()` needs to take in the state that the process should be left in if restarting fails.

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -411,6 +411,12 @@ impl Kernel {
                                 }
+                                Syscall::EXIT { mode } => {
+                                    if config::CONFIG.trace_syscalls {
+                                        debug!("[{:?}] notify_panic({:?})", appid, mode);
```
Given that we would like to use this to enable apps to signal exits in cases other than panics, I think this debug statement should change.

torfmaster replied:
Sure, I fixed that.

bradjc commented:
For the syscall "exit", I think it should do three things:
- Unconditionally it tells the kernel to no longer run the process without resetting its state (including memory, grants, etc.) first.
- It requires the app to pass an exit code to the kernel, as commonly happens on other systems.
- It allows the app to ask the kernel that it try re-running the process (after resetting it). It is, of course, up to the kernel whether the kernel wants to actually do this.

torfmaster commented:
> Marking as "blocked" because I don't think we can merge this into master since it changes the syscall interface. We will likely need to start a tock2.0 branch and merge it there.

Is that necessary? The changes are non-breaking (apps will not notice) so merging them would be in principle okay. But I'm fine with blocking this for now.

gendx commented:
I'd also say that adding a non-essential syscall should rather be a minor revision (Tock 1.5) than requiring bumping the major version number. On a related note, there have been quite a bit of changes since Tock 1.4 so it would also be useful to have a 1.5 version incorporating these before we do the breaking changes of Tock 2.0.

jrvanwhy commented:
> What I'd really like to see is a reliable way to get the panic message from userspace so that if the console driver itself isn't working we can still get the message. This may be bring up specific but robust error reporting is important.

I sort-of implemented this when I added the "application panic" alert code to the [LowLevelDebug](https://github.com/tock/tock/blob/master/doc/syscalls/00008_low_level_debug.md) capsule. It provides a way for an application with no working console driver to report a panic. The downside is that `LowLevelDebug` is only meant to be a debugging tool, not a part of a production Tock kernel, and will not always be present.

In Rust (I'm not familiar with `libtock-c`'s error handling), a panic does not represent an undefined state. The main concern with `panic!()` calling into arbitrary routines is accidentally recursing into a routine that wasn't designed to be reentrant — but doing so won't produce undefined behavior, just unexpected results. At the time I implemented the `LowLevelDebug` panic reporting, I expected all `panic!()` calls to fall into one of the following two cases:

1. Code generation, relocation, or another low-level thing is broken and *nothing* except handwritten inline assembly can be expected to work. In this case, `panic!()` making it to libtock-rs' panic handler is already kinda lucky. This should never happen in a production application, only during runtime development (which I've spent a lot of time on).

2. The language runtime itself is in a perfectly working and sane state and `panic!()` can print a message using Console.

When I did that, I didn't think through case 2. The problem is when the runtime is working perfectly (case 2) but the application has already started a console write. In that case, `panic!()` may need to wait until the console is available to print its message. Currently, it has no way to do so without spinning or invoking callbacks.

bradjc commented:
I still think that having one giant change that we can work towards is better than slowly changing the syscall interface. That is partly motivated by documentation; if this PR is merged the Tock documentation would be inconsistent. I find it is easier to review and update the documentation as one larger pass.

ppannuto commented:
https://github.com/tock/tock/pull/1624#pullrequestreview-365375722
> In general I don't like that this currently leads to a situation where handling of faults in libtock-c apps is very different than the handling of panics in libtock-rs apps.

I think this is actually okay -- C does not really have an equivalent notion of a panic, C++ sort of does with exceptions and could leverage a similar mechanisms to Rust.

---

I'm less convinced this needs to hold till 2.0 rather than 1.5, but I do think we need to fully flesh out the exit/management syscall rather than implementing it piecemeal.

---

I think debug stuff @labbott is getting at is interesting, and suggests that this should perhaps be more of a "process management" syscall than such a narrowly scoped  "exit" syscall.

---

I think Brad summarized the major todos this needs as just an exit syscall well:  https://github.com/tock/tock/pull/1624#issuecomment-592044628

---

To move this forward, I think I'd like to see the exit syscall more completely implemented, as Brad described, and then a discussion of the pros/cons of a more generic process management syscall, and what beyond process stop/restart and core dumps should be part of such a syscall.

bradjc commented:
Can this be rebased and issued to the https://github.com/tock/tock/tree/tock-2.0 branch?

torfmaster commented:
I cannot change the target branch so I would have to close the PR.

hudson-ayers commented:
I changed the base, you should be able to rebase it now

bradjc commented on kernel/src/process.rs:
```diff
@@ -655,6 +662,18 @@ impl<C: Chip> ProcessType for Process<'a, C> {
+                self.terminate(State::Exited);
+            }
+            ExitMode::Restart => {
+                static ALWAYS_RESTART: AlwaysRestart = AlwaysRestart::new();
```
I don't think that apps should be able to circumvent the kernel settings for restart. The `restart_app` function probably shouldn't take a policy as an argument (the policy should already be attached to the process somehow).

Also, we should open a new PR to move the restart code to its own function, and then the 2.0 change will be a little more compact.

bradjc commented:
`restart_app()` is now a function, which should simplify the implementation of this.

torfmaster commented:
I rebased on the tock-2.0 branch and hopefully addressed the remaining review comments.

torfmaster commented:
> Unfortunately this now seems to have reimplemented a lot of the changes from #1841. I brought the tock-2.0 branch even with master, so now that shows as a merge conflict. I think this needs to be rebased on tock-2.0 again, and changed to use the restart() function introduced in that PR

As this is already the fourth time I rebased this I'll wait until I have some certainty that this (very tiny) change will land in tock-2.0 soon. As a volunteer I simply don't have the time.

Two comments:

 * `proccess.rs` is _huge_ (ca. 5 time as large as I usually permit file to be) and this is causing a lot of merge conflicts. It should be split.
 * I would discuss using a different release model than release branches (release from master with feature gates, versioned apis, ...)

bradjc commented:
Plan is the tock-2.0 team will incorporate this into their full 2.0 change.

bradjc commented:
@phil-levis @lschuermann Can we merge this into the 2.0 branch?

lschuermann commented:
@bradjc On a first glance this appears to be largely compatible with the changes already made (and to be made) on the `tock-2.0-dev` branch. However, as the name implies, we still in the development stage. Personally I'd appreciate it to tackle one issue at a time on the branch and only changes which would break the ABI.

Given that this only adds something and should not break existing apps, I guess both is fine: merging into master either now (which would generate some conflicts later on) or preferably after merging the breaking changes of `tock-2.0-dev` into master.

hudson-ayers commented:
Earlier in this thread there is reasoning about why we did not want to gradually change the system call interface, but instead have all changes to the interface occur as one (when merging 2.0 with master), even for "non-breaking" changes that just add system calls.

lschuermann commented:
> Earlier in this thread there is reasoning about why we did not want to gradually change the system call interface

Thanks for pointing this out! I agree with the reasoning, my motivation was to try keeping the diff small. :smile: 
As I said, I'm fine either way. Will put this on my 2.0 todo list then and discuss with @phil-levis on a strategy when & how to integrate this.

We'd probably wait until the other system calls are fairly stable prior to integrating these changes. From the previous comments I suspect @torfmaster isn't particularly motivated to retarget this PR to the `tock-2.0-dev` branch as soon as we've reached a certain level of confidence in the implementation there?

phil-levis commented:
This was discussed on the core call today; there was a consensus that we should rope `exit` into the Tock 2.0 updates to the system call API/ABI. The current proposed approach is to make it a variant of yield (`yield-exit`).

bradjc commented:
Exit has been merged into the 2.0 branch, so I think we can close this PR. We'll get this done soon!
